### PR TITLE
manila: Switch manila-api to run under WSGI

### PR DESCRIPTION
**Pull request by s-t-e-v-e-n-k**
The current release of manila no longer supports running the API service
directly while using SSL, and instead requires the use of WSGI.

---

I'm not certain this correct for HA, it deploys, but only one of the clone sets runs successfully because it binds to the VIP.

### CODE REVIEW

**Comment by s-t-e-v-e-n-k** [comment]
I have confirmed that manila-api prior to this change is deployed as a clone set, so it shouldn't be making use of the VIP ...

File: chef/cookbooks/manila/recipes/api.rb [hunk 1]
```diff
@@ -131,8 +133,50 @@
 
 crowbar_pacemaker_sync_mark "create-manila_register"
 
-use_crowbar_pacemaker_service = ha_enabled && node[:pacemaker][:clone_stateless_services]
+service "manila-api" do
+  service_name "manila-api"
```
Comment (vuntz) [inline chef/cookbooks/manila/recipes/api.rb]: This should be:

```ruby
"openstack-manila-#{params[:name]}" if %w(rhel suse).include? node[:platform_family]
```
(according to manila/definitions/manila_service.rb)

File: chef/cookbooks/manila/recipes/api.rb [hunk 2]
```diff
@@ -131,8 +133,50 @@
...
+  end
+end
+
+apache_site "manila-api-admin.conf" do
+  enable true
+end
+
+crowbar_openstack_wsgi "WSGI entry for manila-api-public" do
+  bind_host my_public_host
+  bind_port manila_port
```
Comment (vuntz) [inline chef/cookbooks/manila/recipes/api.rb]: Why do we have two WSGI services? Why not just one listening on all IP addresses?

File: chef/cookbooks/manila/recipes/common.rb [hunk 3]
```diff
@@ -14,6 +14,7 @@
 #
 
 package "openstack-manila"
+package "openstack-manila-api"
```
Comment (vuntz) [inline chef/cookbooks/manila/recipes/common.rb]: I doubt this change is correct; manila-share nodes shouldn't have this.

File: chef/cookbooks/manila/recipes/controller_ha.rb [hunk 4]
```diff
@@ -46,9 +48,25 @@
...
+  transaction_objects.push(objects)
+
+  order_name = "o-cl-apache2-manila-api"
+  pacemaker_order order_name do
+    ordering "cl-apache2 cl-manila-api"
+    score "Mandatory"
+    action :update
+    only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
+  end
+  transaction_objects << "pacemaker_order[#{order_name}]"
```
Comment (vuntz) [inline chef/cookbooks/manila/recipes/controller_ha.rb]: I don't think you need the dummy resource, unless you have anything else that has a dependency on it (like was the case for keystone). I'd drop that part completely.

File: chef/cookbooks/manila/recipes/api.rb (hunk 2 above)
Comment (s-t-e-v-e-n-k) [reply chef/cookbooks/manila/recipes/api.rb]: Sorry, I was following the Keystone example, conveniently forgetting that public and admin for manila are on the same port.

File: chef/cookbooks/manila/recipes/api.rb [hunk 5]
```diff
@@ -131,8 +133,50 @@
...
+  action [:disable, :stop]
+  # allow to fail here because there may not be a service
+  ignore_failure true
+end
+
+ssl_enabled = node[:manila][:api][:protocol] == "https"
+
+crowbar_openstack_wsgi "WSGI entry for manila-api-admin" do
+  bind_host my_admin_host
+  bind_port manila_port
```
Comment (rhafer) [inline chef/cookbooks/manila/recipes/api.rb]: I think the bind_host/bind_port settings need to be different depending on ha/no-ha setup (to allow haproxy to listen on the main manila port). See https://github.com/crowbar/crowbar-openstack/blob/master/chef/cookbooks/nova/recipes/placement_api.rb#L93 for an example how we do it in other places.

**Comment by rhafer** [comment]
@s-t-e-v-e-n-k I've pushed some additional changes, which should address most of the review comments. With that I could get it to successfully apply in an HA deployment. I didn't really test it yet, neither did I try a SSL deployment.

**Comment by jsuchome** [comment]
@s-t-e-v-e-n-k please keep the git messages from ralf's commits next time, thanks

**Comment by rhafer** [comment]
> please keep the git messages from ralf's commits next time, thanks

I think he just didn't notice my commit before force-pushing. As some of my changes are missing. I'll rebase my changes and push again.

**Comment by rhafer** [comment]
@s-t-e-v-e-n-k  I've added a few fixes. Feel free to squash them into your commit.

File: chef/cookbooks/manila/recipes/api.rb [hunk 6]
```diff
@@ -131,8 +135,40 @@
...
+  ssl_certfile node[:manila][:ssl][:certfile]
+  ssl_keyfile node[:manila][:ssl][:keyfile]
+  if node[:manila][:ssl][:cert_required]
+    ssl_cacert node[:manila][:ssl][:ca_certs]
+  end
+end
 
-manila_service "api" do
-  use_pacemaker_provider use_crowbar_pacemaker_service
+apache_site "manila-api.conf" do
```
Comment (dirkmueller) [inline chef/cookbooks/manila/recipes/api.rb]: maybe use openstack-manila-api.conf here?

File: chef/cookbooks/manila/recipes/api.rb [hunk 7]
```diff
@@ -131,8 +135,40 @@
...
+  supports status: true, restart: true
+  action [:disable, :stop]
+  # allow to fail here because there may not be a service
+  ignore_failure true
+end
+
+ssl_enabled = node[:manila][:api][:protocol] == "https"
+
+if node[:manila][:ha][:enabled]
+  admin_address = Chef::Recipe::Barclamp::Inventory.get_network_by_type(node, "admin").address
```
Comment (vuntz) [inline chef/cookbooks/manila/recipes/api.rb]: For the record: no need for the `Chef::Recipe::` prefix here.

File: chef/cookbooks/manila/recipes/api.rb [hunk 8]
```diff
@@ -131,8 +135,40 @@
...
+  manila_bind_host = admin_address
+  manila_port = node[:manila][:ha][:ports][:api]
+else
+  manila_bind_host = "0.0.0.0"
+  manila_port = manila_port
```
Comment (vuntz) [inline chef/cookbooks/manila/recipes/api.rb]: Interesting line ;-)

Comment (rhafer) [reply chef/cookbooks/manila/recipes/api.rb]: Heh, I guess that was a copy'n past oversight ...

File: chef/cookbooks/manila/recipes/api.rb (hunk 6 above)
Comment (rhafer) [reply chef/cookbooks/manila/recipes/api.rb]: It's seems most other service settle for leaving the `openstack` away.

**Pull request merged**